### PR TITLE
[Packetbeat] Add support for af_packet fanout group

### LINKED ISSUE #35451: [Packetbeat] Add config option for AF_PACKET fanout group
**Describe the enhancement:**

Add a configuration to set an AF_PACKET fanout group (`uint16`). If two different Packetbeat processes use the same group ID, they'll "share" packets between them. By utilizing `PACKET_FANOUT_HASH` mode Linux will split traffic between each member process. Flows with the same hash are always routed to the same process. Each Packetbeat process must be configured with the same af_packet device settings.

Packetbeat may need to use `PACKET_FANOUT_FLAG_DEFRAG` to get consistent hashing in the case of fragmented packets.

- https://pkg.go.dev/github.com/google/gopacket/afpacket#TPacket.SetFanout

- https://www.man7.org/linux/man-pages/man7/packet.7.html

**Describe a specific use case for the enhancement or feature:**

This will allow multiple Packetbeat processes to run in parallel and split the work.

### PR DESCRIPTION
**Pull request by andrewkroh**
## What does this PR do?

To scale processing across multiple Packetbeat processes, a fanout group
identifier can be specified. When `fanout_group` is used the Linux kernel splits
packets across Packetbeat instances in the same group by using a flow hash. It
computes the flow hash modulo with the number of Packetbeat processes in order
to consistently route flows to the same Packetbeat instance.

The value must be between 0 and 65535. By default, no value is set.

This is only available on Linux and requires using `type: af_packet`. Each process
must be running in same network namespace. All processes must use the same
interface settings. You must take responsibility for running multiple instances of
Packetbeat.

Closes #35451

## Why is it important?

Gives advanced users a means of scaling packet capture across Packetbeat processes on Linux.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test

1. Create config file as shown below with `fanout_group`.
2. Launch multiple instances of Packetbeat. Each with their own `path.data` directory.

    ```
    ./packetbeat -e -c ./packetbeat.dev.yml --path.data /tmp/packetbeat1 &
    ./packetbeat -e -c ./packetbeat.dev.yml --path.data /tmp/packetbeat2 &
    ```
3. Generate some traffic.
4. Query Elasticsearch. Verify that no `network.community_id` is associated with more than one `agent.id`. The means that each flow is consistently routed to one Packetbeat instance.

Screenshot 2023-05-12 at 19 29 42

```yaml
---
packetbeat.interfaces:
- device: any
  fanout_group: 1
  type: af_packet

packetbeat.interfaces.internal_networks:
  - private

packetbeat.protocols:
- type: dns
  ports: [53]

- type: http
  ports: [80, 8080, 8000, 5000, 8002]

- type: tls
  ports: [443]

output.elasticsearch:
  hosts:
    - https://host.docker.internal:9200
  ssl.verification_mode: none
  username: elastic
  password: changeme
  allow_older_versions: true
```

## Related issues

- Closes #35451

### CODE REVIEW

File: packetbeat/_meta/config/beat.reference.yml.tmpl [hunk 1]
```diff
@@ -44,6 +44,20 @@ packetbeat.interfaces.internal_networks:
...
+# packets across Packetbeat instances in the same group by using a flow hash. It
+# computes the flow hash modulo with the number of Packetbeat processes in order
+# to consistently route flows to the same Packetbeat instance.
+#
+# The value must be between 0 and 65535. By default, no value is set.
+#
+# This is only available on Linux and requires using `type: af_packet`. Each process
+# must be running in same network namespace. All processes must use the same
+# interface settings. You must take responsibility for running multiple instances
+# of Packetbeat.
```
Comment (efd6) [inline packetbeat/_meta/config/beat.reference.yml.tmpl:58]: I'm wondering about the relative merits of running this as n×processes v n×sniffers in one process.

If we give the config an n rather than an ID, in `sniffer.New` the sniffer construction loop can do n sniffer constructions per device (fnv(devname+pid) for fanout id?) and then the user does not need to coordinate the process starts, nor make sure there is no id collision — though there should be a warning for this). The question is whether the OS scheduler will do a better job at distributing work than the runtime scheduler. Also, " The same should work for multiple TPacket objects within the same process." [in the docs](https://pkg.go.dev/github.com/google/gopacket/afpacket#TPacket.SetFanout) is probably less affirmative than I really like, though the mechanism is via fd, so there is no reason why it is not true.

Comment (taylor-swanson) [reply packetbeat/_meta/config/beat.reference.yml.tmpl:58]: I'm torn on this as well. From a customer-deployment perspective, keeping everything contained in one process would be much simpler. The points that Dan brings up are valid, though.

Comment (andrewkroh) [reply packetbeat/_meta/config/beat.reference.yml.tmpl:58]: To start I totally think we should give consideration to allowing users to run concurrent sniffers on the same interface. My thinking with this change is that it does not prevent us from allowing this in the future. I'm glad you raised this because it will make sure we are thinking about how the config would work so that we don't do something that causes grief when we want to support n×sniffers.

> relative merits of running this as n×processes v n×sniffers in one process

When I was testing this I tried the approach of running multiple sniffers via config, but Packetbeat has a [restriction](https://github.com/elastic/beats/blob/1496ba72cc1d6f5ebfdb2b183132b8f55c4f66b7/packetbeat/config/config.go#L84) that prevents this.

```
packetbeat.interfaces:
- device: any
  fanout_group: 1
  type: af_packet
- device: any
  fanout_group: 1
  type: af_packet
```

> The question is whether the OS scheduler will do a better job at distributing work than the runtime scheduler.

I don't know the answer and it would be nice to have a recommendation. By allowing n×processes we are giving some flexibility for controlling process level settings like CPU affinity. Probably the runtime would do just as good a job, but I don't know.

> fnv(devname+pid) for fanout id?

I think an explicit fanout ID is better. If the user is operating any other AF_PACKET processes (like Zeek) then by having a predicable fanout ID they can ensure there are no collisions.


------

In summary, I think we can do an independent enhancement to support the n×sniffers model. The config would offer a `concurrency` option.

```
packetbeat.interfaces:
- device: any
  concurrency: 4 # Number of concurrent tasks to create for reading packets from the interface.
  fanout_group: 20001
  type: af_packet
```

Comment (efd6) [reply packetbeat/_meta/config/beat.reference.yml.tmpl:58]: > > relative merits of running this as n×processes v n×sniffers in one process
> 
> When I was testing this I tried the approach of running multiple sniffers via config, but Packetbeat has a [restriction](https://github.com/elastic/beats/blob/1496ba72cc1d6f5ebfdb2b183132b8f55c4f66b7/packetbeat/config/config.go#L84) that prevents this.
> 
> ```
> packetbeat.interfaces:
> - device: any
>   fanout_group: 1
>   type: af_packet
> - device: any
>   fanout_group: 1
>   type: af_packet
> ```

I was thinking rather of having a numeric config something like

```
packetbeat.interfaces:
- device: any
  type: af_packet
  fanout: 2
  fanout_group_id: 1
```

This then constructs the fanout group and applies the policy to it.

> I don't know the answer and it would be nice to have a recommendation.

I don't have an answer.

**Review by efd6 — APPROVED** [review]

**Review by taylor-swanson — APPROVED** [review]
I think this is a great start. As Andrew mentioned, we can revisit the n x sniffers model in a separate, focused PR.
